QccPack-0.61-1/src/libQccPackHYP/utils/man1/rawtoicb.1
.TH RAWTOICB 1 "QCCPACK" ""
.SH NAME
rawtoicb \- converts raw file to image cube
.SH SYNOPSIS
.B rawtoicb
.RB "[\|" \-s "\|]"
.RB "[\|" \-bpv
.IB bits_per_voxel "\|]"
.RB "[\|" \-bsq "\|]"
.RB "[\|" \-bil "\|]"
.RB "[\|" \-bip "\|]"
.RB "[\|" \-bend "\|]"
.RB "[\|" \-lend "\|]"
.I num_rows
.I num_cols
.I num_frames
.I rawfile
.I icbfile
.SH OPTIONS
.TP
.BI \-s
Indicates raw file contains signed values.
Default: raw file contains unsigned values.
.TP
.BI \-bpv
Indicates the number of bits per each voxel (sample) value.
Default is 16.
.TP
.BI \-bsq
Indicates raw file is stored in band-sequential (BSQ) order. This is
the default.
.TP
.BI \-bil
Indicates raw file is stored in band-interleaved-by-line (BIL) order.
.TP
.BI \-bip
Indicates raw file is stored in band-interleaved-by-pixel (BIP) order.
.TP
.BI \-bend
Indicates raw file contains 16-bit integers stored in
big-endian format. This is the default.
.TP
.BI \-lend
Indicates raw file contains 16-bit integers stored in
little-endian format.
.SH DESCRIPTION
.LP
.B rawtoicb
converts a raw file
containing a 3D array of integer samples (voxels) into a
.BR ICB -format
file. The size of the raw file must be given as
.IR num_rows ,
.IR num_cols ,
and
.IR num_frames
.RI ( num_frames
= number of spectral bands).
.LP
The option
.BR \-bpv
specifies the number of bits that each voxel value contains.
.IR bits_per_voxel
must be between 1 and 32. If
.IR bits_per_voxel
is <= 8, then each sample is stored in
.IR rawfile
as the least significant
bits in a single byte; otherwise, if
.IR bits_per_voxel
is <= 16, then each sample is stored in
.IR rawfile
as the least significant bits of two consecutive bytes (the default);
otherwise, if
.IR bits_per_voxel
is <= 32, then each sample is stored in
.IR rawfile
as the least significant bits of four consecutive bytes.
The option
.BR \-s
indicates that the raw file contains signed values;
unsigned is assumed if the
.BR \-s
option is absent.
The options
.BR \-bend
and
.BR \-lend
indicate whether the raw file is in
big-endian or little-endian
format, respectively.
If neither is specified,
big-endian is assumed.
.LP
The options
.BR \-bsq ,
.BR \-bil ,
and
.BR \-bip
indicate the scanning order of the rawfile;
see
.BR QccHYPRaw (3).
If none of
.BR \-bsq ,
.BR \-bil ,
or
.BR \-bip
are present,
BSQ is assumed.
.SH "SEE ALSO"
.BR icbtoraw (1),
.BR QccHYPRaw (3),
.BR QccPackHYP (3),
.BR QccPack (3)

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackHYP/utils/rawtoicb.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "rawtoicb.h"


#define USG_STRING "[-s %:] [-bpv %d:bits_per_voxel] [-bsq %:] [-bil %:] [-bip %:] [-bend %:] [-lend %:] %d:num_rows %d:num_cols %d:num_frames %s:rawfile %s:icbfile"

QccString RawFile;

int NumRows;
int NumCols;
int NumFrames;

QccIMGImageCube ImageCube;

int Signed = 0;

int BitsPerVoxel = 16;

int Format;
int FormatBSQSpecified = 0;
int FormatBILSpecified = 0;
int FormatBIPSpecified = 0;

int Endian;
int BigEndianSpecified = 0;
int LittleEndianSpecified = 0;


int main(int argc, char *argv[])
{
  QccInit(argc, argv);
  
  QccIMGImageCubeInitialize(&ImageCube);

  if (QccParseParameters(argc, argv,
			 USG_STRING,
                         &Signed,
                         &BitsPerVoxel,
                         &FormatBSQSpecified,
                         &FormatBILSpecified,
                         &FormatBIPSpecified,
                         &BigEndianSpecified,
                         &LittleEndianSpecified,
                         &NumRows,
                         &NumCols,
                         &NumFrames,
                         RawFile,
                         ImageCube.filename))
    QccErrorExit();
  
  if ((BitsPerVoxel < 1) ||
      (BitsPerVoxel > 32))
    {
      QccErrorAddMessage("%s: bpv must be between 1 and 32",
                         argv[0]);
      QccErrorExit();
    }
      
  if (FormatBSQSpecified + FormatBILSpecified + FormatBIPSpecified > 1)
    {
      QccErrorAddMessage("%s: Only one of -bsq, -bil, or -bip can be specified",
                         argv[0]);
      QccErrorExit();
    }

  if (FormatBSQSpecified)
    Format = QCCHYP_RAWFORMAT_BSQ;
  else
    if (FormatBILSpecified)
      Format = QCCHYP_RAWFORMAT_BIL;
    else
      if (FormatBIPSpecified)
        Format = QCCHYP_RAWFORMAT_BIP;
      else
        Format = QCCHYP_RAWFORMAT_BSQ;

  if (BigEndianSpecified + LittleEndianSpecified > 1)
    {
      QccErrorAddMessage("%s: Only one of -be or -le can be specified",
                         argv[0]);
      QccErrorExit();
    }

  if (BigEndianSpecified)
    Endian = QCCHYP_RAWENDIAN_BIG;
  else
    if (LittleEndianSpecified)
      Endian = QCCHYP_RAWENDIAN_LITTLE;
    else
      Endian = QCCHYP_RAWENDIAN_BIG;

  ImageCube.num_rows = NumRows;
  ImageCube.num_cols = NumCols;
  ImageCube.num_frames = NumFrames;
  if (QccIMGImageCubeAlloc(&ImageCube))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageCubeAlloc()",
                         argv[0]);
      QccErrorExit();
    }

  if (QccHYPRawRead3D(RawFile, &ImageCube,
                      BitsPerVoxel, Signed, Format, Endian))
    {
      QccErrorAddMessage("%s: Error calling QccHYPRawRead3D()",
                         argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageCubeSetMaxMin(&ImageCube))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageCubeSetMaxMin()",
                         argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageCubeWrite(&ImageCube))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageCubeWrite()",
                         argv[0]);
      QccErrorExit();
    }

  QccIMGImageCubeFree(&ImageCube);

  QccExit;
}
